commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala

```diff
@@ -237,16 +240,65 @@ final class JsonObjectInput(reader: JsonReader, options: JsonOptions, callback:
 
   def hasNext: Boolean = !end
 
-  def nextField(): JsonStringFieldInput = {
+  private def nextFieldName(): String = {
     reader.skipWs()
     val fieldName = reader.parseString()
     reader.skipWs()
     reader.pass(':')
-    new JsonStringFieldInput(fieldName, reader, options, this)
+    fieldName
   }
 
+  def nextField(): JsonStringFieldInput =
+    new JsonStringFieldInput(nextFieldName(), reader, options, this)
+
   def afterElement(): Unit =
     prepareForNext(first = false)
+
+  override def peekField(name: String): Opt[FieldInput] = {
+    def peekFieldInput(name: String, idx: Int): JsonStringFieldInput = {
+      val forkedReader = new JsonReader(reader.json).setup(_.reset(idx))
+      new JsonStringFieldInput(name, forkedReader, options, AfterElementNothing)
+    }
+
+    def skipToNextFieldName(): Int = {
+      new JsonStringInput(reader).skip()
+      reader.skipWs()
+      if (reader.isNext('}')) -1
+      else {
+        reader.pass(',')
+        reader.index
+      }
+    }
+
+    val alreadyPeeked = peekedFields.opt.flatMap(_.getOpt(name).map(idx => peekFieldInput(name, idx)))
+    alreadyPeeked orElse {
+      val savedIdx = reader.index
+      try {
```

Comment:
this try/catch/finally is not sufficiently tested, everything passes with this omitted

Reply:
Well it's rather paranoic. I'm not sure if something can throw exceptions in there.

Reply:
Seems odd that this part is the only one which gave you this feeling :D I'd probably opt for not doing it and finding out, but it's your call ;)

Reply:
This part temporarily modifies state of `reader` and must restore after it's done.
Something like this automatically makes me write a `try`-`finally`.